src/hooks/useUSDCPrice.ts

```diff
@@ -17,43 +18,64 @@ export const STABLECOIN_AMOUNT_OUT: { [chainId: number]: CurrencyAmount<Token> }
   [SupportedChainId.POLYGON]: CurrencyAmount.fromRawAmount(USDC_POLYGON, 10_000e6),
 }
 
+const v2USDCPriceCache = new Map<Currency, Price<Currency, Token>>()
+const v3USDCPriceCache = new Map<Currency, Price<Currency, Token>>()
+
 /**
  * Returns the price in USDC of the input currency
  * @param currency currency to compute the USDC price of
  */
 export default function useUSDCPrice(currency?: Currency): Price<Currency, Token> | undefined {
+  // Use a cache per block to avoid recomputing USDC price.
+  // USDC price computation involves finding the best route for a currency, and is computationally expensive and blocking.
+  const block = useBlockNumber()
+  useEffect(() => {
+    v2USDCPriceCache.clear()
+    v3USDCPriceCache.clear()
+  }, [block])
+
   const chainId = currency?.chainId
 
   const amountOut = chainId ? STABLECOIN_AMOUNT_OUT[chainId] : undefined
   const stablecoin = amountOut?.currency
 
   // TODO(#2808): remove dependency on useBestV2Trade
-  const v2USDCTrade = useBestV2Trade(TradeType.EXACT_OUTPUT, amountOut, currency, {
-    maxHops: 2,
-  })
-  const v3USDCTrade = useClientSideV3Trade(TradeType.EXACT_OUTPUT, amountOut, currency)
+  const v2Currency = currency && v2USDCPriceCache.has(currency) ? undefined : currency
+  const v2USDCTrade = useBestV2Trade(TradeType.EXACT_OUTPUT, amountOut, v2Currency, { maxHops: 2 })
```

Comment:
a better fix is probably to not rely on the v2 router at all. The client side v3 trade should already be cached for a given polling interval.

Reply:
The v3 trade may be cached, but running the hook still takes 1-3ms, and this hook is invoked in multiple components to compute usdc prices, so this easily causes dropped frames. Also, until the trade is cached, it is computed in each of the components that calls the hook, so it's not just a dropped frame, but ends up being a few 100ms of non-interactivity.

I think a best solution is to do both:
- Just use the v3 trade
- Cache the v3 trade from the first caller, for usdc price only (as this hook is the only entry point AFAIK that is re-used across both app and widget)